[Security Solution][Expandable flyout] - replace advanced settings with feature flag

PhilippeOberti opened the pull request:
## Summary

The goal of this PR is to simplify the current expandable flyout advanced settings and feature flag setup. We currently have 1 advanced settings and 5 feature flags around the usage of the expandable flyout in Security Solution. The advanced settings and all the feature flags are `true` by default.

This PR consolidates the 5 feature flags and the advanced settings into a single feature flag. The new flag is called `expandableFlyoutDisabled` and is set to `false` by default. That way customers can turn the usage of the expandable flyout within Security Solution with one change in their `kibana.yml` file, should the need arise.

Here are the main changes:
- remove the `securitySolution:enableExpandableFlyout`  advanced settings, so that users cannot revert to the old flyout through the UI
- combine the `expandableEventFlyoutEnabled`, `expandableFlyoutInCreateRuleEnabled`, `expandableTimelineFlyoutEnabled`, `newUserDetailsFlyout` and `newHostDetailsFlyout` feature flags. All these flags were `true` by default so we're not expecting customers to change these values.

### TO DO

- [x] fix unit test
- [x] fix e2e tests

https://github.com/elastic/security-team/issues/9360

PhilippeOberti committed:
[Security Solution][Expandable flyout] - replace advanced settings with feature flag

PhilippeOberti commented:
/ci

PhilippeOberti commented:
buildkite test this

PhilippeOberti commented:
/ci

PhilippeOberti commented:
buildkite test this

PhilippeOberti commented:
buildkite test this

PhilippeOberti commented:
buildkite test this

PhilippeOberti commented:
buildkite test this

elasticmachine commented:
Pinging @elastic/security-threat-hunting-investigations (Team:Threat Hunting:Investigations)

PhilippeOberti commented:
## Files by Code Owner

### elastic/appex-sharedux

* packages/kbn-management/settings/setting_ids/index.ts

### elastic/kibana-core

* src/plugins/kibana_usage_collection/server/collectors/management/schema.ts
* src/plugins/kibana_usage_collection/server/collectors/management/types.ts
* src/plugins/telemetry/schema/oss_plugins.json

### elastic/kibana-localization

* x-pack/plugins/translations/translations/fr-FR.json
* x-pack/plugins/translations/translations/ja-JP.json
* x-pack/plugins/translations/translations/zh-CN.json

### elastic/kibana-management

* packages/kbn-management/settings/setting_ids/index.ts
* packages/serverless/settings/security_project/index.ts

### elastic/kibana-telemetry

* src/plugins/telemetry/schema/oss_plugins.json

### elastic/security-defend-workflows

* x-pack/plugins/security_solution/public/management/cypress/e2e/policy/policy_details.cy.ts
* x-pack/plugins/security_solution/public/management/cypress/e2e/policy/policy_experimental_features_disabled.cy.ts
* x-pack/plugins/security_solution/public/management/cypress/e2e/policy/policy_list.cy.ts
* x-pack/plugins/security_solution/public/management/cypress/e2e/response_actions/isolate.cy.ts
* x-pack/plugins/security_solution/public/management/cypress/e2e/response_actions/isolate_mocked_data.cy.ts
* x-pack/plugins/security_solution/public/management/cypress/e2e/sentinelone/isolate.cy.ts
* x-pack/plugins/security_solution/public/management/cypress/tasks/common.ts

### elastic/security-detection-engine

* x-pack/plugins/security_solution/public/detections/pages/alerts/alert_details_redirect.test.tsx
* x-pack/plugins/security_solution/public/detections/pages/alerts/alert_details_redirect.tsx
* x-pack/plugins/security_solution/public/detections/pages/alerts/utils.ts
* x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/detection_alerts/enrichments/alert_threat_enrichments.cy.ts

### elastic/security-engineering-productivity

* x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/detection_alerts/enrichments/alert_threat_enrichments.cy.ts
* x-pack/test/security_solution_cypress/cypress/e2e/entity_analytics/enrichments.cy.ts
* x-pack/test/security_solution_cypress/cypress/e2e/explore/guided_onboarding/tour.cy.ts
* x-pack/test/security_solution_cypress/cypress/e2e/investigations/alerts/alerts_details.cy.ts
* x-pack/test/security_solution_cypress/cypress/e2e/investigations/alerts/investigate_in_timeline.cy.ts
* x-pack/test/security_solution_cypress/cypress/screens/alerts_details.ts
* x-pack/test/security_solution_cypress/cypress/screens/expandable_flyout/alert_details_right_panel_table_tab.ts
* x-pack/test/security_solution_cypress/cypress/tasks/alerts_details.ts
* x-pack/test/security_solution_cypress/cypress/tasks/api_calls/kibana_advanced_settings.ts

### elastic/security-entity-analytics

* x-pack/test/security_solution_cypress/cypress/e2e/entity_analytics/enrichments.cy.ts

### elastic/security-solution

* packages/serverless/settings/security_project/index.ts
* x-pack/plugins/security_solution/common/constants.ts
* x-pack/plugins/security_solution/common/experimental_features.ts
* x-pack/plugins/security_solution/public/cases/pages/index.tsx
* x-pack/plugins/security_solution/public/common/components/control_columns/row_action/index.test.tsx
* x-pack/plugins/security_solution/public/common/components/control_columns/row_action/index.tsx
* x-pack/plugins/security_solution/public/common/hooks/flyout/use_init_flyout_url_param.ts
* x-pack/plugins/security_solution/public/common/hooks/flyout/use_sync_flyout_url_param.ts
* x-pack/plugins/security_solution/public/detections/pages/alerts/alert_details_redirect.test.tsx
* x-pack/plugins/security_solution/public/detections/pages/alerts/alert_details_redirect.tsx
* x-pack/plugins/security_solution/public/detections/pages/alerts/utils.ts
* x-pack/plugins/security_solution/public/explore/users/containers/users/observed_details/index.tsx
* x-pack/plugins/security_solution/public/flyout/document_details/right/hooks/use_flyout_is_expandable.ts
* x-pack/plugins/security_solution/public/management/cypress/e2e/policy/policy_details.cy.ts
* x-pack/plugins/security_solution/public/management/cypress/e2e/policy/policy_experimental_features_disabled.cy.ts
* x-pack/plugins/security_solution/public/management/cypress/e2e/policy/policy_list.cy.ts
* x-pack/plugins/security_solution/public/management/cypress/e2e/response_actions/isolate.cy.ts
* x-pack/plugins/security_solution/public/management/cypress/e2e/response_actions/isolate_mocked_data.cy.ts
* x-pack/plugins/security_solution/public/management/cypress/e2e/sentinelone/isolate.cy.ts
* x-pack/plugins/security_solution/public/management/cypress/tasks/common.ts
* x-pack/plugins/security_solution/public/timelines/components/open_timeline/note_previews/index.tsx
* x-pack/plugins/security_solution/public/timelines/components/side_panel/hooks/use_detail_panel.test.tsx
* x-pack/plugins/security_solution/public/timelines/components/side_panel/hooks/use_detail_panel.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/body/events/stateful_event.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/body/index.test.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/body/renderers/host_name.test.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/body/renderers/host_name.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/body/renderers/user_name.test.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/body/renderers/user_name.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/unified_components/data_table/index.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/unified_components/hooks/use_unified_timeline_expandable_flyout.test.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/unified_components/hooks/use_unified_timeline_expandable_flyout.tsx
* x-pack/plugins/security_solution/server/ui_settings.ts

### elastic/security-threat-hunting-explore

* x-pack/plugins/security_solution/public/cases/pages/index.tsx
* x-pack/plugins/security_solution/public/explore/users/containers/users/observed_details/index.tsx
* x-pack/test/security_solution_cypress/cypress/e2e/explore/guided_onboarding/tour.cy.ts

### elastic/security-threat-hunting-investigations

* x-pack/plugins/security_solution/public/flyout/document_details/right/hooks/use_flyout_is_expandable.ts
* x-pack/plugins/security_solution/public/timelines/components/open_timeline/note_previews/index.tsx
* x-pack/plugins/security_solution/public/timelines/components/side_panel/hooks/use_detail_panel.test.tsx
* x-pack/plugins/security_solution/public/timelines/components/side_panel/hooks/use_detail_panel.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/body/events/stateful_event.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/body/index.test.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/body/renderers/host_name.test.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/body/renderers/host_name.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/body/renderers/user_name.test.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/body/renderers/user_name.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/unified_components/data_table/index.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/unified_components/hooks/use_unified_timeline_expandable_flyout.test.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/unified_components/hooks/use_unified_timeline_expandable_flyout.tsx
* x-pack/test/security_solution_cypress/cypress/e2e/investigations/alerts/alerts_details.cy.ts
* x-pack/test/security_solution_cypress/cypress/e2e/investigations/alerts/investigate_in_timeline.cy.ts
* x-pack/test/security_solution_cypress/cypress/screens/expandable_flyout/alert_details_right_panel_table_tab.ts

rylnd commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/detection_alerts/enrichments/alert_threat_enrichments.cy.ts:
```diff
@@ -71,39 +69,37 @@ describe('Threat Match Enrichment', { tags: ['@ess', '@serverless', '@skipInServ
     });
   });
 
-  it('Displays persisted enrichments on the JSON view', () => {
```
The addition of the `Table tab` test is good, but could you explain why the JSON view coverage was removed from cypress?

PhilippeOberti replied:
hey @rylnd, I can explain for sure! I tried for 2+ hours to fix the JSON test, but the issue is the new flyout uses the `JsonCodeEditor` from the `unified_doc_viewer` which is different from the old flyout that was using a simple `EuiCodeBlock`. The code block was allowing us to easily retrieve the content and test against it. The `JsonCodeEditor` does not allow to do that.
The only way I kinda found to get it to work was to Cypress scroll to the correct position and select the row, but this is extremely flaky and we should never do that.

My thought was, for this test we don't really care if the value is showing in the JSON tab. The purpose of this test is not to verify that the flyout JSON tab shows the correct value. We should trust that the JSON tab and its components are rendering the document correctly (and we already have tests for that on the expandable flyout section). What we care about if the fact that the document has the correct value. The easiest way I found to do this was by filtering the table and verifying that the expected value was there.

Is that a wrong assumption?

rylnd reviewed (approved):
Detection engine changes overall LGTM.

I had one question about a perceived loss in test coverage, but I'm approving on the assumption that it was a conscious decision. If not, it should be easy to add back the removed test.

christineweng reviewed (approved):
Desk tested with the flag on and off and LGTM! Thanks for cleaning up all the cypress tests!

PhilippeOberti committed:
skip test

PhilippeOberti commented on x-pack/plugins/security_solution/public/management/cypress/e2e/response_actions/isolate_mocked_data.cy.ts:
```diff
@@ -201,7 +197,8 @@ describe('Isolate command', { tags: ['@ess', '@serverless', '@brokenInServerless
 
-  describe('from Cases', () => {
+  // TODO re-enable when https://github.com/elastic/security-team/issues/9625 is merged
+  describe.skip('from Cases', () => {
```
this test highlighted a bug with how the expandable flyout is set up to work in Cases. As the expandable flyout has been enabled since `8.10`, this PR doesn't introduce any new issues, but just reveals the fact that the Cypress tests were behind, by still using the old flyout.

We'll take a look at [the ticket](https://github.com/elastic/security-team/issues/9625) to fix the issue shortly, and will re-enable the tests!

gergoabraham commented on x-pack/plugins/security_solution/public/management/cypress/e2e/response_actions/isolate_mocked_data.cy.ts:
```diff
@@ -92,9 +92,6 @@ describe('Isolate command', { tags: ['@ess', '@serverless', '@brokenInServerless
   describe.skip(
```
just fyi, created ticket for our team to re-enable this: https://github.com/elastic/security-team/issues/9626

gergoabraham reviewed (approved):
@elastic/security-defend-workflows related changes look good! thanks for cleaning up the tests! 👐

angorayc commented on x-pack/test/security_solution_cypress/cypress/e2e/explore/guided_onboarding/tour.cy.ts:
```diff
@@ -30,12 +29,12 @@ import { visit } from '../../../tasks/navigation';
 
-describe('Guided onboarding tour', { tags: ['@ess'] }, () => {
+// re-enable this when https://github.com/elastic/kibana/pull/181442 is merged
+describe.skip('Guided onboarding tour', { tags: ['@ess'] }, () => {
```
Is this test failing?

PhilippeOberti replied:
no, I had skipped it because my branch was opened at the time it was failing and you were working on the fix. It's unskipped in my local, and working!
I'm fixing a couple of other comments then will push a commit. No changes to this file anymore!

PhilippeOberti replied:
@angorayc the change was reverted in this last commit https://github.com/elastic/kibana/pull/184169/commits/7f601451a5169cdb9b502089502369f426c12b5a!

PhilippeOberti committed:
PR comments
- unskip tour cypress test
- update investigate in timeline cypress test
- fix test title

PhilippeOberti commented on x-pack/test/security_solution_cypress/cypress/e2e/explore/guided_onboarding/tour.cy.ts:
```diff
@@ -30,12 +29,12 @@ import { visit } from '../../../tasks/navigation';
 import { deleteAlertsAndRules } from '../../../tasks/api_calls/common';
 
-describe('Guided onboarding tour', { tags: ['@ess'] }, () => {
+// re-enable this when https://github.com/elastic/kibana/pull/181442 is merged
```
good point, reverted the change in https://github.com/elastic/kibana/pull/184169/commits/7f601451a5169cdb9b502089502369f426c12b5a

PhilippeOberti commented on x-pack/test/security_solution_cypress/cypress/e2e/investigations/alerts/investigate_in_timeline.cy.ts:
```diff
@@ -32,78 +31,95 @@ import {
+      ftrConfig: {
+        kbnServerArgs: [
+          `--xpack.securitySolution.enableExperimental=${JSON.stringify([
+            'expandableFlyoutDisabled',
```
yeah, I was being lazy... I added some basic tests (here https://github.com/elastic/kibana/pull/184169/commits/7f601451a5169cdb9b502089502369f426c12b5a) that we should probably improve a bit when we remove this entire old flyout codebase!

machadoum reviewed (approved):
EA changes LGTM!
